contracts/actions/utils/TokenizedVaultAdapter.sol

```diff
@@ -0,0 +1,130 @@
+        params.from = _parseParamAddr(params.from, _paramMapping[3], _subData, _returnValues);
+        params.to = _parseParamAddr(params.to, _paramMapping[4], _subData, _returnValues);
+        params.operationId = OperationId(_parseParamUint(uint8(params.operationId), _paramMapping[5], _subData, _returnValues));
+
+        (bytes memory logData, uint256 returnAmount) = _executeOperation(params);
+        emit ActionEvent("TokenizedVaultAdapter", logData);
+        return bytes32(returnAmount);
+    }
+
+    function executeActionDirect(bytes memory _callData) public payable virtual override {
+        Params memory params = parseInputs(_callData);
+        (bytes memory logData, ) = _executeOperation(params);
+        logger.logActionDirectEvent("TokenizedVaultAdapter", logData);
+    }
+
+    function _executeOperation(Params memory _params) internal returns (bytes memory logData, uint256) {
+        IERC4626 vault = IERC4626(_params.vaultAddress);
+
+        if (_params.from == address(0)) {
+            _params.from = address(this);
+        }
+
+        if (_params.to == address(0)) {
+            _params.to = address(this);
+        }
+
+        if (_params.operationId == OperationId.REDEEM) {
+            if (_params.amount == type(uint256).max) {
+                _params.amount = vault.balanceOf(_params.from);
+            }
+            uint256 assetsWithdrawn = vault.redeem(_params.amount, _params.to, _params.from);
+            if (assetsWithdrawn < _params.minOutOrMaxIn) revert TokenizedVaultSlippageHit();
+            logData = abi.encode(_params, assetsWithdrawn);
+            return (logData, assetsWithdrawn);
+        }
+
+        if (_params.operationId == OperationId.WITHDRAW) {
+            uint256 sharesBurned = vault.withdraw(_params.amount, _params.to, _params.from);
+            if (sharesBurned > _params.minOutOrMaxIn) revert TokenizedVaultSlippageHit();
+            logData = abi.encode(_params, sharesBurned);
+            return (logData, sharesBurned);
+        }
+
+        address assetAddress = vault.asset();
+
+        if (_params.operationId == OperationId.DEPOSIT) {
+            _params.amount = assetAddress.pullTokensIfNeeded(_params.from, _params.amount);
+            assetAddress.approveToken(address(vault), _params.amount);
+
+            uint256 sharesMinted = vault.deposit(_params.amount, _params.to);
+            if (sharesMinted < _params.minOutOrMaxIn) revert TokenizedVaultSlippageHit();
+            logData = abi.encode(_params, sharesMinted);
+            return (logData, sharesMinted);
+        }
+
+        if (_params.operationId == OperationId.MINT) {
+            uint256 assetsToDeposit = vault.previewMint(_params.amount);
+            if (assetsToDeposit > _params.minOutOrMaxIn) revert TokenizedVaultSlippageHit();
+
+            _params.minOutOrMaxIn = assetAddress.pullTokensIfNeeded(_params.from, assetsToDeposit);
```

Comment:
Don't reuse the variable `_params.minOutOrMaxIn` for `uint256 pulledAssetAmount` makes it harder to follow the logic while not saving much on gas